umh-core/pkg/service/filesystem/filesystem.go

```diff
@@ -364,10 +503,90 @@ func (s *DefaultService) WriteFile(ctx context.Context, path string, data []byte
 
 // PathExists checks if a path (file or directory) exists.
 func (s *DefaultService) PathExists(ctx context.Context, path string) (bool, error) {
+	start := time.Now()
+
 	if err := s.checkContext(ctx); err != nil {
 		return false, err
 	}
 
+	// Cache all paths with 1-second TTL for simplicity
+	// This provides good balance between freshness and performance
+	const cacheTTL = 1 * time.Second
+
+	// Check cache
+	s.pathCache.mu.RLock()
+
+	if cached, ok := s.pathCache.cache[path]; ok && time.Now().Before(cached.expiry) {
+		s.pathCache.mu.RUnlock()
+		// Cache hit - record as cached operation
+		metrics.RecordFilesystemOp("PathExists", path, true, time.Since(start))
+
+		return cached.exists, cached.err
+	}
+
+	s.pathCache.mu.RUnlock()
+
+	// Not in cache or expired - perform actual check
+	exists, err := s.pathExistsUncached(ctx, path)
+	// Only cache positive results (when path exists) with 1-second TTL
+	// Don't cache negative results as paths might be created immediately after
+	if err == nil && exists {
+		s.pathCache.mu.Lock()
+		s.pathCache.cache[path] = &CachedPath{
+			exists: exists,
+			err:    err,
+			expiry: time.Now().Add(cacheTTL),
+		}
+		s.pathCache.mu.Unlock()
+	}
+
+	metrics.RecordFilesystemOp("PathExists", path, false, time.Since(start))
+
+	return exists, err
+}
+
+// invalidatePathCache invalidates cache entries for a path and all subpaths.
+func (s *DefaultService) invalidatePathCache(path string) {
+	// Invalidate path existence cache
+	s.pathCache.mu.Lock()
+	// Delete exact path match
+	delete(s.pathCache.cache, path)
+	// Delete any cached paths that are under this path (for RemoveAll)
+	pathWithSlash := path + "/"
```

Comment:
It's probably better to only conditionally add this.
E.g only if it is missing from path.

Reply:
Or even trimming "/" from path at the very start and then adding it.